src/nvme/mi.c

```diff
@@ -582,12 +582,39 @@ int nvme_mi_admin_xfer(nvme_mi_ctrl_t ctrl,
 	}
 
 	/* bidirectional not permitted (see DLEN definition) */
-	if (req_data_size && *resp_data_size) {
+	int direction = admin_req->opcode & 0x3;
+	if (direction == NVME_DATA_TFR_BIDIRECTIONAL || (req_data_size && *resp_data_size))
+	{
+		nvme_msg(ctrl->ep->root, LOG_ERR,
+				 "nvme_mi_admin_xfer doesn't support bidirectional commands\n");
 		errno = EINVAL;
 		return -1;
 	}
-
-	if (!*resp_data_size && resp_data_offset) {
+	else if (direction == NVME_DATA_TFR_HOST_TO_CTRL && *resp_data_size)
+	{
+		nvme_msg(ctrl->ep->root, LOG_ERR,
+				 "nvme_mi_admin_xfer doesn't support response data on WRITE commands\n");
+		errno = EINVAL;
+		return -1;
+	}
+	else if (direction == NVME_DATA_TFR_CTRL_TO_HOST && req_data_size)
+	{
+		nvme_msg(ctrl->ep->root, LOG_ERR,
+				 "nvme_mi_admin_xfer doesn't support request data on READ commands\n");
+		errno = EINVAL;
+		return -1;
+	}
+	else if (resp_data_offset && (direction != NVME_DATA_TFR_CTRL_TO_HOST || !*resp_data_size))
+	{
+		nvme_msg(ctrl->ep->root, LOG_ERR,
+				 "nvme_mi_admin_xfer invalid offset value\n");
+		errno = EINVAL;
+		return -1;
+	}
+	else if (direction == NVME_DATA_TFR_NO_DATA_TFR && req_data_size && *resp_data_size)
+	{
+		nvme_msg(ctrl->ep->root, LOG_ERR,
+				 "nvme_mi_admin_xfer doesn't support data on NONE_DATA_TRANSFER commands\n");
```

Comment:
You have these conditionals on the direction here, and then another set below to set up the request/response fields.

I would suggest simplifying this logic by having a single `switch (direction) { ... }` statement, and checking for the errors, plus setting up the req & resp fields in each `case`.

But if you don't decide to use a `switch`: since all of the statements return, you don't need to make this a chain of `else if`; each check can be its own independent statement. This makes it slightly easier to reason about the logic.